drivers: pwm: stm32: add pwm capture support

TilmannUnte opened the pull request:
Extends STM32 PWM driver to support capturing pulse width, period,
or both.

Fixes #39394

Signed-off-by: Tilmann Unte <unte@es-augsburg.de>

gmarull reviewed (dismissed):
Thanks for this addition, however, this PR seems to mix the changes with a few code style tweaks, making the review process difficult. Code style changes are ok as far as they are contained in their own commit, but not mixed with a feature addition. The overlay file for l475 should also be part of a separate commit.

TilmannUnte committed:
tests: drivers: pwm: stm32 board for loopback test

Add overlay file for disco_l475_iot1 board to pwm_loopback test.

The file is based on the existing overlays for twr_ke18f and frdm_k64f.

The new configuration uses both a 16-Bit timer with active prescaler
as well as a 32-Bit timer with deactivated prescaler for coverage.

It is assumed that other STM32 boards will behave like the
disco_l475_iot1. Only this board was available for testing.

Signed-off-by: Tilmann Unte <unte@es-augsburg.de>

TilmannUnte committed:
drivers: pwm: stm32: add pwm capture support

Extends STM32 PWM driver to support capturing pulse width, period,
or both.

The approach used is based on the PWM Input sections in reference
manual RM0351 by ST in the chapters on timers.
The LL library by ST is used for all extensions to the driver.

Only the disco_l475_iot1 board was available for testing on hardware,
using tests/drivers/pwm/pwm_loopback.
Since timers are a generic component, it is assumed that other STM32
boards will behave the same

Fixes #39394

Signed-off-by: Tilmann Unte <unte@es-augsburg.de>

ABOSTM reviewed (changes requested):
@TilmannUnte, Thanks for this PR.
This is a nice job and this feature was awaited.
Some inline comments

TilmannUnte commented:
> this link https://github.com/multiwii/baseflight/blob/master/src/drv_pwm.c maybe helpful

I had not seen this before. Unfortunately it does not help much as a reference, because it uses the HAL library, whereas the Zephyr driver exclusively uses LL.

I personally prefer LL, because it has a close relationship to the hardware registers defined in the reference manuals.

TilmannUnte commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -28,10 +28,25 @@ LOG_MODULE_REGISTER(pwm_stm32, CONFIG_PWM_LOG_LEVEL);
 #define IS_TIM_32B_COUNTER_INSTANCE(INSTANCE) (0)
 #endif
 
+struct pwm_stm32_capture_data {
+	pwm_capture_callback_handler_t callback;
+	void *user_data;
+	uint32_t period;
+	uint32_t pulse;
+	uint32_t overflows;
+	uint8_t skip_irq;
+	bool capture_period;
+	bool capture_pulse;
+	bool continuous;
+};
```
Done!

TilmannUnte commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,6 +293,240 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+{
+
+	/*
+	 * From the reference manual RM0351:
+	 * The PWM input mode can be used only with the TIMx_CH1/TIMx_CH2
+	 * signals due to the fact that only TI1FP1 and TI2FP2 are connected to
+	 * the slave mode controller.
+	 */
```
I improved the comment for clarity following your suggestions.

TilmannUnte commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -241,6 +271,10 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 		oc_init.CompareValue = pulse_cycles;
 		oc_init.OCPolarity = get_polarity(flags);
 
+		LL_TIM_SetSlaveMode(cfg->timer, LL_TIM_SLAVEMODE_DISABLED);
+		LL_TIM_SetTriggerInput(cfg->timer, LL_TIM_TS_ITR0);
+		LL_TIM_DisableMasterSlaveMode(cfg->timer);
+
```
I decided to use both

TilmannUnte commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,6 +293,243 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+		return -ENOTSUP;
+	}
+
+	cpt->callback = cb;
```
Yeah, seems like the better way. I also followed your other suggestions related to this one.

TilmannUnte commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -224,6 +247,13 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 	}
 
+#ifdef CONFIG_PWM_CAPTURE
+	if (LL_TIM_IsEnabledIT_CC1(cfg->timer)) {
```
Must've somehow missed that. I also added checks for specific channel numbers.

TilmannUnte commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,6 +293,243 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+	} else {
+		cpt->skip_irq--;
+	}
+	LL_TIM_ClearFlag_CC1(cfg->timer);
+	LL_TIM_ClearFlag_CC2(cfg->timer);
```
That wouldn't be good, I added the check.

TilmannUnte commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,6 +293,243 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+			}
+		}
+	} else {
+		cpt->skip_irq--;
```
I'm clearing the update flag seperately now.

skip_irq is meant to only decrement when a capture occurs. So I included that check.
The update flag only starts to have relevance once the skipped irqs are over.

TilmannUnte commented:
I've worked on the requested changes, but will continue to test locally

ABOSTM reviewed (approved):
LGTM

ABOSTM commented:
BTW, I tested on nucleo_f207zg, and it works fine.

erwango commented on tests/drivers/pwm/pwm_loopback/boards/disco_l475_iot1.overlay:
```diff
@@ -0,0 +1,52 @@
+	status = "okay";
+	pwm3: pwm {
+		status = "okay";
+		st,prescaler = <255>;
```
Should be sync'ed with https://github.com/zephyrproject-rtos/zephyr/pull/39628.
This won't trigger any failure, but we should avoid having both implementations in tree once #39628 is merged

TilmannUnte replied:
I worked out the minor nitpicks. Awaiting #39628 now, thank you for pointing it out to me.

erwango replied:
@TilmannUnte https://github.com/zephyrproject-rtos/zephyr/pull/39628 has just been merged.

TilmannUnte replied:
All done!

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -331,9 +625,28 @@ static int pwm_stm32_init(const struct device *dev)
+	irq_enable(DT_IRQN(DT_PARENT(DT_DRV_INST(index))));                    \
+}
+#define CAPTURE_INIT(index)                                                    \
+.irq_config_func = pwm_stm32_irq_config_func_##index
```
Nitpick: indentation

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -331,9 +625,28 @@ static int pwm_stm32_init(const struct device *dev)
+{                                                                              \
+	IRQ_CONNECT(DT_IRQN(DT_PARENT(DT_DRV_INST(index))),                    \
+			DT_IRQ(DT_PARENT(DT_DRV_INST(index)), priority),       \
+	pwm_stm32_isr, DEVICE_DT_INST_GET(index), 0);                          \
```
Nitpick: indentation

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -331,9 +625,28 @@ static int pwm_stm32_init(const struct device *dev)
 
+#ifdef CONFIG_PWM_CAPTURE
+	cfg->irq_config_func(dev);
+#endif /* CONFIG_PWM_CAPTURE */
```
Break a line

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -259,6 +303,251 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
+	LL_TIM_EnableIT_UPDATE(cfg->timer);
+	LL_TIM_CC_EnableChannel(cfg->timer, LL_TIM_CHANNEL_CH1);
+	LL_TIM_CC_EnableChannel(cfg->timer, LL_TIM_CHANNEL_CH2);
+	return 0;
```
Break a line

erwango reviewed (changes requested):
Thanks for this work, this brings a long awaited feature.

A couple of nitpicks otherwise LGTM.

Main point will be to synch with https://github.com/zephyrproject-rtos/zephyr/pull/39628
(which we'll try to get merged quickly first)

erwango commented:
No related CI failure: tests/kernel/lifo/lifo_usage/kernel.lifo.usage in qemu_cortex_a53_smp:tests/kernel/lifo/lifo_usage/kernel.lifo

TilmannUnte commented:
> No related CI failure: tests/kernel/lifo/lifo_usage/kernel.lifo.usage in qemu_cortex_a53_smp:tests/kernel/lifo/lifo_usage/kernel.lifo

Anything I can/need to do here?

erwango commented:
> > No related CI failure: tests/kernel/lifo/lifo_usage/kernel.lifo.usage in qemu_cortex_a53_smp:tests/kernel/lifo/lifo_usage/kernel.lifo
> 
> Anything I can/need to do here?

You can have a try to rebase on very latest master, hopefully the issue got fixed.
